File: ios/app/Sources/RemoteStore.swift
import Foundation
import Security

/// The one remembered remote VellumFE server (host, port, pairing token)
/// for the Remote login tab, stored as a single Keychain item — same trust
/// posture as the master password key in `CryptoKeys.swift`: device-bound,
/// never in backups. Mirrors the Android shell's `RemoteStore.kt`.
enum RemoteStore {
    struct Target: Codable, Equatable {
        var host: String
        var port: Int
        /// Pairing token for that PC's web server; empty when the user
        /// paired without one (the remote page prompts instead).
        var token: String
    }

    private static let base: [String: Any] = [
        kSecClass as String: kSecClassGenericPassword,
        kSecAttrService as String: "dev.vellumfe.remote-server",
        kSecAttrAccount as String: "vellum-remote",
    ]

    static func load() -> Target? {
        var query = base
        query[kSecReturnData as String] = true
        var item: CFTypeRef?
        guard SecItemCopyMatching(query as CFDictionary, &item) == errSecSuccess,
              let data = item as? Data,
              let target = try? JSONDecoder().decode(Target.self, from: data)
        else { return nil }
        return target
    }

    static func save(_ target: Target) {
        guard let data = try? JSONEncoder().encode(target) else { return }
        SecItemDelete(base as CFDictionary)
        var add = base
        add[kSecValueData as String] = data
        // Same accessibility choice as the master key: usable from a cold
        // launch once the phone has been unlocked, never restored to other
        // hardware (the token only pairs with a PC this device could reach).
        add[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
        let status = SecItemAdd(add as CFDictionary, nil)
        if status != errSecSuccess {
            NSLog("VellumShell: saving remote server failed: \(status)")
        }
    }

    static func forget() {
        SecItemDelete(base as CFDictionary)
    }
}

File: android/app/src/main/java/dev/vellumfe/RemoteStore.kt
package dev.vellumfe

import android.content.Context
import android.util.Log
import org.json.JSONObject
import java.io.File

/**
 * The one remembered remote VellumFE server (host, port, pairing token)
 * for the Remote login tab, sealed with the Keystore wrap key from
 * [CryptoKeys] — same trust posture as the master password key: on-device
 * only, unreadable without this device's Keystore. Mirrors the iOS shell's
 * `RemoteStore.swift`.
 */
object RemoteStore {
    private const val TAG = "VellumShell"
    private const val FILE = "remote.bin"

    data class Target(
        val host: String,
        val port: Int,
        /** Pairing token for that PC's web server; empty when the user
         * paired without one (the remote page prompts instead). */
        val token: String,
    )

    fun load(context: Context): Target? {
        val file = File(context.filesDir, FILE)
        if (!file.exists()) return null
        return try {
            val json = JSONObject(String(CryptoKeys.openBlob(file.readBytes()), Charsets.UTF_8))
            Target(
                host = json.getString("host"),
                port = json.getInt("port"),
                token = json.optString("token"),
            )
        } catch (e: Exception) {
            Log.w(TAG, "saved remote server unreadable; forgetting it: $e")
            file.delete()
            null
        }
    }

    fun save(context: Context, target: Target) {
        try {
            val json = JSONObject()
                .put("host", target.host)
                .put("port", target.port)
                .put("token", target.token)
            File(context.filesDir, FILE)
                .writeBytes(CryptoKeys.sealBlob(json.toString().toByteArray(Charsets.UTF_8)))
        } catch (e: Exception) {
            Log.w(TAG, "saving remote server failed: $e")
        }
    }

    fun forget(context: Context) {
        File(context.filesDir, FILE).delete()
    }
}
